lib/linklist.c

```diff
@@ -139,8 +157,8 @@ void listnode_add_sort(struct list *list, void *val)
 
 	assert(val != NULL);
 
-	new = listnode_new();
-	new->data = val;
+	new = listnode_new(list, val);
```

Comment:
Good Q. And you are right -
1. In the first case (LINKLIST_FLAG_NODE_MEM_BY_APP) the assignment (val = new->data) is needed
2. In the second case (!LINKLIST_FLAG_NODE_MEM_BY_APP) the assignment (val = new->data) is NOT needed i.e. a redundant operation.

I could have added another check i.e. -
if (list->flags & LINKLIST_FLAG_NODE_MEM_BY_APP) 
       val = new->data;

In terms of performance both are more or less the same. I can change it to that if it will be less confusing.